Bug 1147157 - position:fixed elements break layout on orientation change

sleroux opened the pull request:
So this is kind of a gross hack but it works. It looks like there's an issue with WKWebView where it doesn't invalidate it's drawing area on orientation change. However, it does invalidate it whenever the content area is scrolled (see https://github.com/WebKit/webkit/blob/master/Source/WebKit2/UIProcess/API/Cocoa/WKWebView.mm#L1430). I added a small method as an extension to WKWebView that will nudge the content area by a single pixel which forces the content area to invalidate and added this to the traitCollectionChange callback in BrowserViewController.

thebnich commented:
Yuck, this is a gross hack, but nice find!

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -126,6 +126,11 @@ class BrowserViewController: UIViewController {
     override func traitCollectionDidChange(previousTraitCollection: UITraitCollection?) {
```
I wonder why we're using this instead of `viewWillTransitionToSize`? If we used `viewWillTransitionToSize` instead, we could even do something like this to "undo" the hack:

```
    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        coordinator.animateAlongsideTransition({ animation in
            if let tab = self.tabManager.selectedTab {
                // scroll offset by 1
            }
        }, completion: { context in
            if let tab = self.tabManager.selectedTab {
                // reset scroll offset
            }
        })
    }
```

sleroux replied:
Ya this totally works. I like this better since it keeps the end scroll state unchanged. I updated this PR with a commit that adds this.

thebnich commented on Utils/Extensions/WKWebViewExtensions.swift:
```diff
@@ -0,0 +1,19 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Nit: align the `*`s

thebnich commented on Utils/Extensions/WKWebViewExtensions.swift:
```diff
@@ -0,0 +1,19 @@
+    *  performs a device rotation. Since scrolling calls
+    *  _updateVisibleContentRects (https://github.com/WebKit/webkit/blob/master/Source/WebKit2/UIProcess/API/Cocoa/WKWebView.mm#L1430)
+    *  this method nudges the web view's scroll view by a single pixel to force it to invalidate.
+    */
```
Nit: align the `*`s

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -124,9 +124,31 @@ class BrowserViewController: UIViewController {
+        *  performs a device rotation. Since scrolling calls
+        *  _updateVisibleContentRects (https://github.com/WebKit/webkit/blob/master/Source/WebKit2/UIProcess/API/Cocoa/WKWebView.mm#L1430)
+        *  this method nudges the web view's scroll view by a single pixel to force it to invalidate.
+        */
```
Nit: `*` alignment (or just use `//` since this isn't a documentation comment anymore).

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -124,9 +124,31 @@ class BrowserViewController: UIViewController {
+        }, completion: { context in
+            if let tab = self.tabManager.selectedTab {
+                let contentOffset = tab.webView.scrollView.contentOffset
+                tab.webView.scrollView.setContentOffset(CGPoint(x: contentOffset.x, y: contentOffset.y - 1), animated: true)
```
If you want, you could factor out the `if let tab = ...` and the `let contentOffset = ...` to outside the `coordinator.animateAlongsideTransition` so you don't have to duplicate them in each block. (If you did that, I guess you'd also change `contentOffset.y - 1` -> `contentOffset.y` since it would still hold the original value).

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -124,9 +124,31 @@ class BrowserViewController: UIViewController {
     override func traitCollectionDidChange(previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
         updateToolbarState()
         showToolbars(animated: false)
```
Do you think these should be moved too? I don't know much about this API -- what triggers a "trait collection" change?

sleroux replied:
Ya viewWillTransitionSize and traitCollectionDidChange both get called for almost the same reason. viewWillTransitionSize fires when the screen size has changed which happens on rotation and traitCollectionDidChange gets called when the view controller's size class changes. Since the size class is defined based on an abstract concept involving device + screen size, this also gets called when the screen rotates. I think that using the traitCollection callback makes more sense in this case because we're showing the same UI for the iPhone landscape/iPad portrait views which share the same size class but not the same screen dimensions.

sleroux commented:
Fixed nit pitcks and also moved the logic into willTransitionToTraitCollection instead of traitCollectionDidChange since we have a reference to the animation coordinator there so we can combine both the toolbar logic and web view refresh in the same callback.

TrashbinPornoDolske commented:
You can lick my Penis and my golden Eggs, and kiss my Ass son of a bitch with a personality Disorder from the deepest pits of hell from the Depths of Mozilla!

Fuck you and your Mother you autistic son of a bitch, you're going to lick Putin's Eggs, Fucking Faggot Fascist!

The shit out of your Dog, your Cat and your dirty mountain Goat that Faggot Asshole!

Sucking Mozilla Cocks and all together go and suck on Google's Ass full of Shit!

To fuck in your knee you cock suckers! And who the Fuck Australis Chrome clone of shit! Fucking Mozilla gay lesbian homosexual bastards of hell!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

sleroux commented:
Missed a commit in rebasing so I cherry-picked over the patch to a clean branch see #444
